counter.go

```diff
@@ -0,0 +1,127 @@
+package signalfx
+
+import (
+	"math"
+	"sync/atomic"
+	"time"
+)
+
+// A Counter represents a Counter metric (i.e., its internal value is
+// reset to zero by its PostReportHook, although this should not be
+// meaningful to client code).
+type Counter struct {
+	metric     string
+	dimensions map[string]string
+	value      uint64
+}
+
+// NewCounter returns a new Counter with the specified parameters.  It
+// does not copy the dimensions; client code should take care not to
+// modify them in a goroutine-unsafe manner.
+func NewCounter(
+	metric string,
+	dimensions map[string]string,
+	value uint64,
+) *Counter {
+	return &Counter{metric: metric, dimensions: dimensions, value: value}
+}
+
+// Inc increments a Counter's internal state.  It is goroutine-safe.
+func (c *Counter) Inc(delta uint64) uint64 {
+	return atomic.AddUint64(&c.value, delta)
+}
+
+// DataPoint returns a DataPoint reflecting the Counter's internal
+// state.  It does not reset that state, leaving that to
+// PostReportHook.
+func (c *Counter) DataPoint() *DataPoint {
+	value := atomic.LoadUint64(&c.value)
+	if value == 0 || value > math.MaxInt64 {
+		return nil
+	}
+	return &DataPoint{
+		Metric:     c.metric,
+		Timestamp:  time.Now(),
+		Type:       CounterType,
+		Dimensions: c.dimensions,
+		Value:      int64(value),
+	}
+}
+
+// PostReportHook resets a Counter's internal state.  It does this
+// intelligently, by subtracting the successfully-reported value from
+// the state (this way, the Counter may have been incremented in
+// between the create of the reported DataPoint and the reset, and the
+// state will be consistent).  Calling PostReportHook with a negative
+// value will result in a panic: counters may never take negative
+// values.
+func (c *Counter) PostReportHook(v int64) {
+	if v < 0 {
+		panic("negative counter should be impossible")
```

Comment:
Yeah, I agree that normally in Go one would normally return an error and let the caller figure out how to deal with it. Panics should be reserved to dealing with unrecoverable states.

I think that this qualifies as an unrecoverable state: the argument to a HookedMetric's PostReportHook method is the value return in that HookedMetric's DataPoint method; if that's not a valid value for that HookedMetric, something has gone seriously badly awry. I believe that if something's gone that wrong, then the only thing to do is die and let the orchestration system spin up a new process and alert ops that something weird happened.

There's also the issue that it's unclear how Reporter.Report (which is the only caller of PostReportHook) should recover: should it delete the HookedMetric; should it try to create a new one; should it call some caller-provided recovery function to try to Do The Right Thing in an application-specific manner (and really, 'some caller-provided recovery function' is what panic & recover are all about)?

And then there's the issue that the usual use case is not for clients to directly call Reporter.Report, but rather Reporter.RunInBackground: what should it do if a particular HookedMetric is corrupted? Should it try to handle the issue? Should it pass the value on some sort of error channel back to code running in the main module? Should it call some error-recovery function defined by the main module?

I _think_ that these questions are basically impossible to answer, and thus that this is truly a fatal error, and thus—being a fatal error—it is appropriate to die.